### ISSUE TITLE
Randomly connects to Sink with wifictl help

### ISSUE THREAD

**Issue by wosss** [albfan/miraclecast#259]
I've been trying to get Samsung S7  (source) mirroring screen to Ubuntu 16.04.3 (sink): 1.- Whenever I try sinkctl and wifid, phone cant see miraclecast. 2.- But trying sinkctl, wifid and wifictl with command "connect" sometimes it gets mirrored with video and sound perfectly, altought just got it twice or three times. I uploade case 1 and 2 logs with names unsuccessfull.zip and succesfull.zip. Please help.
[unsuccessfull.zip](https://github.com/albfan/miraclecast/files/1714908/unsuccessfull.zip)
[successful.zip](https://github.com/albfan/miraclecast/files/1714910/successful.zip)

**Comment by albfan** [comment]
Please read steps to use miraclecast as sink on README, they are basically:

- Stop your wifi (miraclecast will create a wpa_supplicant with Wifi Direct enabled)
- Start miracle-wifid (daemon that talk with wifi interface through dbus)
- Start miracle-sinkctl (controller that make miraclecast to act as a sink)
- Search for sinks from your source (these steps differ from device to device)
- Share screen (devices do the handshaking to create a direct network between both and miracle-sinkctl communicates through miracle-wifid and spawns a stream player when source starts to share its screen)

miracle-wifictl is just a controller for still unimplemented miraclecast as source #4. Do not makes sense it makes any difference here

On what you need help exactly?

**Comment by wosss** [comment]
Thanks for reply, the search for sinks steps is not working for me, that's the part i'm asking for help. Tried three different Android phones, in fact sink can see peers, but peers can't see sink. On my first post successful.zip belongs to case 2, but trying to get it working can take up to 50 times or more.

**Comment by albfan** [comment]
Seems with miracle-wifictl you start devices handshaking from sink. After that peers can see sink and work correctly.

You need your peer devices to initiate p2p_scan. (Samsung has his own screen sharing protocol, maybe it is not 100% miracast compatible, really don't know

I think I need to add the datetime to logs to be able to mix them and see when all the things had happened.

So after connect with miracle-wifictl your phone (S7 edge) sees miraclecast?

**Comment by wosss** [comment]
Not always, trying with miracle-wifictl i have another 3 cases:
1.- A Wi-Fi Direct connection with the following miracle-sinkctl error: 
  `NOTICE: HUP/ERR on socket
ERROR: cannot connect to remote host (111): Conexión rehusada (sink_connected() in ../../../src/ctl/ctl-sink.c:398)`
2.- No connection with following miracle-wifictl error:
`[FAIL] Peer: 00:00:00:00:00:00@4 Reason: unknown`
3.- A successfull connection to sink:
![sink](https://user-images.githubusercontent.com/18101245/36130675-f77c5e7a-103b-11e8-91cf-ccb3dc3f266e.png)

**Comment by albfan** [comment]
Seen the phone screen in miraclecast sink is always managed by `miracle-sinkctl`. What I'm trying to triage here is how to make your device to see miraclecast.

As said, miracle-wifictl is not full implemented so it only can launch a p2p-scan

When you use miracle-wifictl it is the one who starts the handshaking. Seems, after that, your device recognizes miraclecast and you can start the screen sharing. Am I right?

**Comment by wosss** [comment]
Yes, when I use  miracle-wifictl, device can see miraclecast, in fact after "connect" command, the phone (it is rooted) shows this debug messages in wpa_cli:
`<3>P2P-DEVICE-FOUND 00:00:00:00:00:00 p2p_dev_addr=00:00:00:00:00:00 pri_dev_type=1-0050F204-1 name='myhostname' config_methods=0x80 dev_capab=0x21 group_capab=0x8 wfd_dev_info=0x00111c4400c8 icon=0x0 prefix='' ss_dev_info=0x0 freq=2462 new=1
<3>P2P-GO-NEG-REQUEST 00:00:00:00:00:00 dev_passwd_id=4 go_intent=7
IFNAME=p2p0`
It'll keep trying until peer got lost:
`IFNAME=p2p0 <3>P2P-BIGDATA-GROUP-OWNER-INTENT p2p_dev_addr=00:00:00:00:00:00 go_intent=7
<3>P2P-GO-NEG-REQUEST 00:00:00:00:00:00 dev_passwd_id=4 go_intent=7
<3>P2P-GO-NEG-REQUEST 00:00:00:00:00:00 dev_passwd_id=4 go_intent=7
<3>P2P-GO-NEG-REQUEST 00:00:00:00:00:00 dev_passwd_id=4 go_intent=7
<3>P2P-DEVICE-LOST p2p_dev_addr=00:00:00:00:00:00`

This is the trial and error part, got a few times a screen share working but most of the times could not share any content.

**Comment by albfan** [comment]
Then we need to know why samsung needs sink to find it before it allows to share content. 

This is from samsung s7 manual:

> Smart View: View your device’s content on the TV screen. You may need to activate the Screen Mirroring feature on the TV depending on the TV you want to connect. If the TV does not support the Screen Mirroring feature, use compatible accessories to enable the feature. Compatible accessories include HomeSync and an AllShare Cast dongle.

We need the specs of his "smart view" (surely an extension of miracast spec) to know how to make it compatible, but I don't own any samsung so cannot develop or test anything.

Maybe you can try if it works on other miracast implementations (like windows 10 or other dongles like easycast or anycast) but you will need to hack on miraclecast to make it work, are you able to do so?

**Comment by wosss** [comment]
Of course, right now I tried Windows 10 with AirServer Miracast Implementation, also tried AnyCast dongle. Both devices appears as sink as soon I press Smart View (Screen Mirroring) icon at S7 Edge Phone. Then content sharing starts fine.

**Comment by albfan** [comment]
If you can capture net traffic on your phone we can see how they do the handshaking

I suppose someone need to take a look at http://developer.samsung.com/tv/develop/extension-libraries/smart-view-sdk/introduction/ and implement the details. I'm sure anycast and Windows 10 have that implemented

Anyway other users report they share they phone screen with miraclecast so don't know exactly what is happening here

**Comment by albfan** [comment]
I have to review the difference between successful and unsuccessful connection as I think there are some messages unhandled by miracast that some devices use for pairing. That could fix lot of problems from several devices to pair with miraclecast
